File: .orkestrel/campaign/docs-parity/instruments/d7/units/sea/d7n-sea-converge/cells.mjs
// Compares every non-`Summary` cell of every table row against the baseline.
// Reads the baseline from stdin-supplied path and the current guide, keys each
// row by its first cell, and reports a key whose leading cells moved. The
// inserted `Shape` column is reported separately, as an addition rather than a
// change. Splits on a pipe not preceded by a backslash.
import { readFileSync } from 'node:fs'

const SPLIT = /(?<!\\)\|/

/**
 * Reads every table row of one guide, keyed by its first cell.
 *
 * @param path - The file to read.
 * @returns Each row key mapped to its cells, the final `Summary` cell dropped.
 */
function readRows(path) {
	const rows = new Map()
	for (const line of readFileSync(path, 'utf8').split('\n')) {
		if (!line.startsWith('| ')) continue
		const cells = line
			.split(SPLIT)
			.slice(1, -1)
			.map((cell) => cell.trim())
		if (cells.length < 3) continue
		if (cells[1].startsWith('---')) continue
		rows.set(cells[0], cells.slice(0, -1))
	}
	return rows
}

const before = readRows(process.argv[2])
const after = readRows(process.argv[3])
const moved = []
const gone = []
const added = []
for (const [key, cells] of before) {
	const now = after.get(key)
	if (now === undefined) {
		gone.push(key)
		continue
	}
	const kept = now.filter((cell, at) => at !== 2 || now.length === cells.length)
	if (kept.join(' | ') !== cells.join(' | ')) {
		moved.push(`${key}: before [${cells.join(' | ')}] after [${kept.join(' | ')}]`)
	}
}
for (const key of after.keys()) if (!before.has(key)) added.push(key)
process.stdout.write(`rows before: ${before.size}, rows after: ${after.size}\n`)
process.stdout.write(`rows missing after: ${gone.length}${gone.length > 0 ? ` (${gone.join(', ')})` : ''}\n`)
process.stdout.write(`rows added: ${added.length}${added.length > 0 ? ` (${added.join(', ')})` : ''}\n`)
process.stdout.write(`non-Summary cells mismatched: ${moved.length}\n`)
for (const line of moved) process.stdout.write(`  ${line}\n`)

File: .orkestrel/campaign/docs-parity/instruments/d7/units/browser/d7n-browser-converge/cells.py
"""Compares every non-Summary cell of the guide against the committed baseline."""
import re, sys

CELL = re.compile(r'(?<!\\)\|')

def rows(path):
	out = {}
	order = []
	for line in open(path).read().split('\n'):
		if not line.startswith('|') or set(line.replace('|', '').replace(' ', '')) <= {'-'}:
			continue
		cells = [c.strip() for c in CELL.split(line)[1:-1]]
		if len(cells) < 2:
			continue
		key = cells[0]
		out.setdefault(key, []).append(cells)
		order.append(key)
	return out, order

base, base_order = rows(sys.argv[1])
now, now_order = rows(sys.argv[2])

print('rows in baseline:', len(base_order), 'rows now:', len(now_order))
missing = [k for k in base_order if k not in now]
added = [k for k in now_order if k not in base]
print('keys missing now:', missing)
print('keys added now:', added)

moved = []
for key, entries in now.items():
	before = base.get(key)
	if before is None:
		continue
	for one, two in zip(entries, before):
		if one[0] != two[0]:
			moved.append((key, 'column 1', two[0], one[0]))
		if len(one) > 1 and len(two) > 1 and one[1] != two[1]:
			moved.append((key, 'column 2', two[1], one[1]))
print('non-Summary cells moved in columns 1 and 2:', len(moved))
for one in moved:
	print('  ', one)
